models/user_heatmap.go

```diff
@@ -0,0 +1,38 @@
+package models
+
+import (
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/util"
+)
+
+// UserHeatmapData represents the data needed to create a heatmap
+type UserHeatmapData struct {
+	Timestamp     util.TimeStamp `json:"timestamp"`
+	Contributions int64          `json:"contributions"`
+}
+
+// GetUserHeatmapDataByUser returns an array of UserHeatmapData
+func GetUserHeatmapDataByUser(user *User) (hdata []*UserHeatmapData, err error) {
+
+	var groupBy string
+	switch {
+	case setting.UseSQLite3:
+		groupBy = "strftime('%Y-%m-%d', created_unix, 'unixepoch')"
+	case setting.UseTiDB:
```

Comment:
`TiDB` is no longer bundled in gitea, and so you can remove this case

Reply:
Done.